sdks/python/apache_beam/pipeline.py

```diff
@@ -525,6 +525,31 @@ def run(self, test_runner_api='AUTO'):
     self.contains_external_transforms = (
         ExternalTransformFinder.contains_external_transforms(self))
 
+    # Finds if RunInference has side inputs enables.
+    # also, checks for the side input window is global and has non default
+    # triggers.
+    run_inference_visitor = RunInferenceVisitor().visit_run_inference(self)
+    self._run_inference_contains_side_input = (
+        run_inference_visitor.contains_run_inference_side_inputs)
+
+    self.run_inference_global_window_non_default_trigger = (
+        run_inference_visitor.contains_global_windows_non_default_trigger)
+
+    if (self._run_inference_contains_side_input and
+        not self._options.view_as(StandardOptions).streaming):
+      raise RuntimeError(
+          "SideInputs to RunInference PTransform is only supported "
```

Comment:
Why?

Reply:
For this side input, the target use cases would reading from unbounded sources or using PeriodicImpulse(which doesn't work as expected in batch mode). 

So by default, the model update side input is only restricted for streaming.

Reply:
Specifically, because batch mode can completely process precondition stages before processing dependent stages, the meaning of pipeline updates breaks down (what does updating to the latest version of the model mean when all inferences are done at the "same time"), and if you do have a streaming source (which our built ins will be), the pipeline can hang forever.

Reply:
In batch, the "latest" version of the model is likely to be the only version of the model, but that doesn't mean it can't be computed lazily. This also avoids having to have two separate codepaths (known at construction time can be "upgraded" to Create + Side Input). 

If you do have a streaming source, the pipeline should automatically run in streaming mode. (Also, IMHO, if just because a mode is more useful in one mode than the other doesn't mean we should prohibit it in the other unless it's actually detrimental.)

Reply:
> > If you do have a streaming source, the pipeline should automatically run in streaming mode.
> 
> AFAIK, this isn't actually true in Python today.

If that's the case, we should definitely fix this :). 

> > if just because a mode is more useful in one mode than the other doesn't mean we should prohibit it in the other unless it's actually detrimental.
> 
> I think its detrimental in the batch use case because understanding the behavior requires understanding runner internals (in this case, how Dataflow/other runners handle stages), and the behavior is very confusing if you don't understand those internals. This is antithetical to one of the purposes of RunInference: taking a hard beam thing and making it easy.

Could you clarify a bit more why understanding runner internals is required for understanding what happens in the batch case (or, in other words, what confusing behavior the user would run into)? I'm not proposing we do away with the "easy" mode when the model is known at compile time (for batch or streaming), rather that we allow its computation to be deferred in both modes if this is explicitly what the user asks for. 

> I want RunInference users to be able to think about Beam as little as possible when building their pipelines, even if it comes at the cost of some of the purity of the unified model.

Yes, for sure. 

> One possible reconciliation here would be to emit a warning in batch mode with an informative error message, or allow users to intentionally enable running in batch mode with a parameter (though neither of those seem like awesome experiences, and I'd still personally favor explicitly disallowing this).

Reply:
>>> If you do have a streaming source, the pipeline should automatically run in streaming mode.

>> AFAIK, this isn't actually true in Python today.

> If that's the case, we should definitely fix this :).

I tend to agree, but IIRC there were concerns about doing this (@tvalentyn thoughts?). It may have just been a matter of us not getting around to doing it yet. I would at least be worried about our existing IOs behaving correctly if we were to change this behavior, I'm not sure how crisp we've been here and I know of at least one (PeriodicImpulse/Sequence) that wasn't correctly marking itself unbounded. That's all solvable though, and probably isn't made worse by the change regardless.

> Could you clarify a bit more why understanding runner internals is required for understanding what happens in the batch case (or, in other words, what confusing behavior the user would run into)? I'm not proposing we do away with the "easy" mode when the model is known at compile time (for batch or streaming), rather that we allow its computation to be deferred in both modes if this is explicitly what the user asks for.

If the following things happened:

1) Pipeline starts with model A
2) 10,000 records arrive for inference + I do some preprocessing
3) Model updates to model B
4) 10,000 records arrive
5) Model updates to model C
6) 1 more records arrives

If I'm in batch mode, all 20,001 records get processed by model C (I actually think this depends on the runner, correct me if I'm wrong). If I'm in streaming mode, model A and model B each process 10,000 records, model C processes 1 record. The streaming case is by far the more useful one, and almost definitely the use case that most people are after with this kind of feature, but the discrepancy here is odd if you don't know how the model works.

With all that said, I can see cases where the batch use case would be nice (e.g. it allows daily pipelines without updating your model manually) - I think the framing of "deferred computation" vs "live model updates" helps there, and I could get behind allowing this. Though, the idea of a single unchanged transform for batch and streaming pipelines still seems fairly unrealistic to me because the boundedness of the side input source would need to be changed to match the boundedness of the main source in most cases to be useful.

My biggest sticking point is still that AFAIK unbounded sources don't automatically put pipelines in streaming mode, which leads to unexplained pipeline hangs in batch mode for this use case. Whether we change that or not, my vote would be to not block this PR on that change.

Reply:
> > > > If you do have a streaming source, the pipeline should automatically run in streaming mode.
> 
> > > AFAIK, this isn't actually true in Python today.
> 
> > If that's the case, we should definitely fix this :).
> 
> I tend to agree, but IIRC there were concerns about doing this (@tvalentyn thoughts?). It may have just been a matter of us not getting around to doing it yet. I would at least be worried about our existing IOs behaving correctly if we were to change this behavior, I'm not sure how crisp we've been here and I know of at least one (PeriodicImpulse/Sequence) that wasn't correctly marking itself unbounded. That's all solvable though, and probably isn't made worse by the change regardless.

Filed https://github.com/apache/beam/issues/25264 ; we can continue discussion there (including if there's any downsides to this change, though I don't see any). If there is undesirable behavior, like hanging, when unbounded sources are used, this is entirely orthogonal to RunInferrence. 

> > Could you clarify a bit more why understanding runner internals is required for understanding what happens in the batch case (or, in other words, what confusing behavior the user would run into)? I'm not proposing we do away with the "easy" mode when the model is known at compile time (for batch or streaming), rather that we allow its computation to be deferred in both modes if this is explicitly what the user asks for.
> 
> If the following things happened:
> 
> 1. Pipeline starts with model A
> 2. 10,000 records arrive for inference + I do some preprocessing
> 3. Model updates to model B
> 4. 10,000 records arrive
> 5. Model updates to model C
> 6. 1 more records arrives
> 
> If I'm in batch mode, all 20,001 records get processed by model C (I actually think this depends on the runner, correct me if I'm wrong). If I'm in streaming mode, model A and model B each process 10,000 records, model C processes 1 record. The streaming case is by far the more useful one, and almost definitely the use case that most people are after with this kind of feature, but the discrepancy here is odd if you don't know how the model works.

One can't have this scenario for batch. In batch, one has 20,001 records and three models. None of the models (or records) are "first" and if anything there is ambiguity which model should be used. (Actually, with AsSingleton, an error will generally be raised if there's more than one.)

Now if the model (and records) are appropriately timestamped (this works in batch), they'll all get the "right" model. Bonus: this makes testing a lot easier. 

> With all that said, I can see cases where the batch use case would be nice (e.g. it allows daily pipelines without updating your model manually) - I think the framing of "deferred computation" vs "live model updates" helps there, and I could get behind allowing this. Though, the idea of a single unchanged transform for batch and streaming pipelines still seems fairly unrealistic to me because the boundedness of the side input source would need to be changed to match the boundedness of the main source in most cases to be useful.

A bounded side input with an unbounded main input is easy to understand, and the other way around can work too (especially if one is using timestamped data) though a bit odd. But we shouldn't fall into the trap that RunInferrence is always going to be at the top level (i.e. at the same level of the Reads)--it may be used in an even higher-level composite for example. 

> My biggest sticking point is still that AFAIK unbounded sources don't automatically put pipelines in streaming mode, which leads to unexplained pipeline hangs in batch mode for this use case. Whether we change that or not, my vote would be to not block this PR on that change.

If using an unbounded side input causes hangs, yes, let's warn on that (ideally as high a level as makes sense, it's not a RunInference-specific issue) until this is fixed. But I'd rather not prohibit side inputs altogether.

Reply:
> Filed https://github.com/apache/beam/issues/25264 ;

Thanks

> None of the models (or records) are "first" and if anything there is ambiguity which model should be used.

I meant first by timestamp. For the model at least (if not the main input), I think this is a reasonable understanding of how the non-Beam person will see the world.

> Now if the model (and records) are appropriately timestamped (this works in batch), they'll all get the "right" model.

I didn't think this was the case, and it alleviates my concerns here.

> If using an unbounded side input causes hangs, yes, let's warn on that (ideally as high a level as makes sense, it's not a RunInference-specific issue) until this is fixed. But I'd rather not prohibit side inputs altogether.

I'm fine with this if Anand is.

Reply:
I had an offline conversation with Danny with this and we both tend to agree on Robert's view.

I imposed the restriction because it would be easier to restrictive at first and then introduce new features(SI for batches) based on the feedback but now I think it is important to follow the beam unified model concept and remove the restriction on the side input.

Thanks